File: The 3n+1 problem/solution.cpp
// @BEGIN_OF_SOURCE_CODE

#include <iostream>

using namespace std;

bool isOdd(int number)
{
  return (number % 2);
}

int getCycleLength(int number) 
{
  if (number <= 1) {
    return 1;
  }

  if (isOdd(number)) {
    return 1 + getCycleLength(number * 3 + 1);
  }
  else {
    return 1 + getCycleLength(number / 2);
  }
}

int getMaxCycleLengthInRange(int i, int j)
{
  int maxCycleLength = 0;
  for (int currentNumber=i; currentNumber<=j; currentNumber++) {
    int cycleLength = getCycleLength(currentNumber);
    if (cycleLength > maxCycleLength) {
      maxCycleLength = cycleLength;
    }
  }
  return maxCycleLength;
}

int main(int argc, char **argv) 
{
  int firstValue;
  int secondValue;

  while (cin >> firstValue >> secondValue) {
    int firstValueOrig = firstValue;
    int seconfValueOrig = secondValue;

    if ( firstValue > secondValue ) swap (firstValue, secondValue);
    
    int maxCycleLength = getMaxCycleLengthInRange(i, j);
    cout << iOrig << " " << jOrig << " " << maxCycleLength << endl;
  }
  return 0;
}

// @END_OF_SOURCE_CODE

File: The 3n+1 problem/solution.swift
#!/usr/bin/env xcrun swift

func isOdd(number: Int) -> Bool
{
  return (number % 2 != 0)
}

func getCycleLength(number: Int) -> Int 
{
  if number <= 1 {
    return 1
  }
 
  if isOdd(number) {
    return 1 + getCycleLength(number * 3 + 1);
  }
  else {
    return 1 + getCycleLength(number / 2);
  }
}

func getMaxCycleLengthInRange(firstValue: Int, secondValue: Int) -> Int 
{
  var maxCycleLength = 0
  for currentNumber in firstValue...secondValue {
    let cycleLength = getCycleLength(currentNumber)
    if cycleLength > maxCycleLength {
      maxCycleLength = cycleLength
    }
  }
  return maxCycleLength
}

func readEntryValues() -> (Int, Int)
{
  let firstValue = readLine(stripNewline: true)
  let secondValue = readLine(stripNewline: true)

  let firstValueInt = Int(firstValue!)!
  let secondValueInt = Int(secondValue!)!

  return (firstValueInt, secondValueInt)
}

func main() 
{
  var (firstValue, secondValue) = readEntryValues() 
  if firstValue > secondValue {
    swap(&firstValue, &secondValue)
  }

  let result = getMaxCycleLengthInRange(firstValue, secondValue: secondValue)

  print(" -- result -- \(result)")
}

main()
